### ISSUE TITLE
Viewer: Incorrect handling of screen IDs in full-screen mode

### ISSUE THREAD

**Issue by dcommander** [TigerVNC/tigervnc#488]
The [description of the SetDesktopSize message](https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#setdesktopsize) says:

"The id field must be preserved upon modification as it determines the difference between a moved screen and a newly created one. The client should make every effort to preserve the fields it does not wish to modify, including any unknown flags bits."

Furthermore, the [description of the ExtendedDesktopSize pseudo-encoding](https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#extendeddesktopsize-pseudo-encoding) says:

"The id field contains an arbitrary value that the server and client can use to map RFB screens to physical screens. The value must be unique in the current set of screens and must be preserved for the lifetime of that RFB screen. New ids are assigned by whichever side creates the screen. An id may be reused if there has been a subsequent update of the screen layout where the id was not used."

To me, this would suggest strongly that the viewer should always retain the screen ID for existing screens when it requests a remote desktop resize.  Doing so clearly informs the server which RANDR output it should modify, rather than requiring the server to guess.  However, the TigerVNC Viewer doesn't do that when switching to full-screen mode.  Instead, it will always create a new screen ID for full-screen mode unless the full-screen viewport is the same size as the non-full-screen viewport.  That is, IMHO, broken, and the [code comments](https://github.com/TigerVNC/tigervnc/blob/master/vncviewer/DesktopWindow.cxx#L940) seem to agree with me on that.

What I would expect is that, when switching from non-full-screen mode to full-screen mode, the viewer will retain the screen ID for existing screens and create a new ID only for any new screens it needs to create.  For instance, if "Enable full-screen mode over all monitors" is checked, the client has multiple monitors, but the server only has one screen configured, then it is perfectly acceptable for the viewer to use a new screen ID for the second screen, but it should use the existing (server-generated) screen ID for the first screen.

**Comment by dcommander** [comment]
Obviously this code has been out there for a while, so any server implementations will have to work around the current behavior, but it would be good to make the viewer behave correctly for precisely the reasons stated in the RFB proto doc:  so the server can tell the difference between a truly new screen and one that has just moved.

**Comment by CendioOssman** [comment]
Well, switching between windowed and full screen isn't exactly "moving" an existing screen, so I wouldn't call this incorrect behaviour. I can agree it's probably sub-optimal though..

**Comment by dcommander** [comment]
You're right.  It isn't moving a screen at all.  It's re-using one of the existing screens, which is exactly why the same ID should be used.  The actual screen on the client doesn't change just because you're suddenly opening a full-screen window on it.  What point is a screen ID if it has no meaningful relationship with an actual screen in either the server-side or client-side display environment?

Furthermore, it's more than just sub-optimal.  Further testing reveals that your viewer implementation actually prevents a proper server-side implementation, because the supposedly random screen IDs your viewer is assigning aren't actually random (you failed to seed the random number generator, so `rand()` always produces the same sequence of numbers.)  Take the following scenario:

1. I start my server implementation with `-geometry 2560x800`.
1. I connect to the server with the TigerVNC Viewer using `vncviewer {host}:{display} -remoteresize -fullscreenallmonitors`, from a dual-screen Mac client (1344x840+0+0, 1344x840+1344+0).
   * This resizes the viewer window to 2560x791, in order to accommodate the Mac menu bar.
   * At the moment, the viewer is using the screen ID that my server sent it.
1. Switch to full-screen mode.
   * This changes the screen layout to 1344x840+0+0, 1344x840+1344+0, with new screen IDs assigned by the viewer.  On my machine, these IDs are always 0x77a4044d and 0x31169898, respectively.
   * The server sees that none of those IDs is known, so it assigns the first one to the existing screen and creates a new screen for the second one.
1. Exit full-screen mode.
   * This changes the screen layout on the server back to 2560x791+0+0 with a disabled secondary screen.
1. Exit the viewer.
1. Reconnect, also using `vncviewer {host}:{display} -remoteresize -fullscreenallmonitors`.
   * Since the server was nice enough to accept the TigerVNC Viewer's new screen IDs from the previous connection, it now sends an ID of 0x77a4044d for the primary screen.  The (currently inactive) secondary screen on the server still has an ID of 0x31169898.
1. Enter full-screen mode.
   * Now the viewer tries to assign "random" IDs again for the "new" screens, but because the RNG isn't seeded, it lands again on ID 0x77a4044d.  Because that ID is already in use by the non-full-screen window and the resolution of the non-full-screen window doesn't match that of the full-screen window, the viewer rejects that ID.  The next two "random" IDs are assigned:  0x31169898 for the primary screen and 0x427c3c55 for the secondary.
   * Because 0x31169898 was assigned to the secondary screen during a previous connection, the server still knows that ID.  When it receives a new EDS request with 0x31169898 = 1344x840+0+0, 0x427c3c55 = 1344x840+1344+0, it tries to do the right thing and assign 0x31169898 to the existing server-side screen with the same ID.  It assigns the unknown ID (0x427c3c55) to the primary screen.
   * The server receives the EDS request as 0x31169898 = 1344x840+0+0, 0x427c3c55 = 1344x840+1344+0, but if it tries to match IDs to existing screens, that results in sending back the EDS message as 0x427c3c55 = 1344x840+1344+0, 0x31169898 = 1344x840+0+0 (with the screen structures reversed.)  Since the viewer doesn't match screen IDs, this effectively causes the screens to be flipped on the client side.

First and foremost, if you're going to retain the current TigerVNC Viewer behavior, then at least seed the RNG so as to avoid duplicate "random" screen IDs in subsequent connections.  That will at least allow a server implementation to work around the TigerVNC Viewer's behavior.  Otherwise, there is no way for a server to distinguish a new screen from one that has moved.  Servers currently have no choice but to forego attempting to match screen IDs at all, which prevents them from handling moved screens in the correct way.

In a nutshell, your implementation requires assuming that the screens are sent in a particular order, but nothing about the protocol requires that.

**Comment by dcommander** [comment]
Update:

Confirmed that the following simple viewer patch works around the issue:

```
--- a/vncviewer/vncviewer.cxx
+++ b/vncviewer/vncviewer.cxx
@@ -31,6 +31,7 @@
 #include <locale.h>
 #include <fcntl.h>
 #include <unistd.h>
+#include <time.h>
 #include <sys/stat.h>
 
 #ifdef WIN32
@@ -460,6 +461,8 @@ int main(int argc, char** argv)
   bindtextdomain(PACKAGE_NAME, LOCALE_DIR);
   textdomain(PACKAGE_NAME);
 
+  srand(time(NULL));
+
   rfb::SecurityClient::setDefaults();
 
   // Write about text to console, still using normal locale codeset
```

**Comment by CendioOssman** [comment]
I'm afraid there has been a misunderstanding here, so perhaps further clarification of the spec is needed.

Monitor id:s are not persistent. If an id isn't used in the current configuration, then it is fair game to use for a new monitor. The extension has no concept of disabled or unused monitor. Servers should forget the id as soon as the monitor is unused. If this wasn't the case then the client would need to be informed of unused monitors when connecting, and not just the active ones.

**Comment by dcommander** [comment]
OK, but your random number generator still needs to be seeded, because otherwise your viewer isn't actually "forgetting" the IDs it assigns.  It's re-assigning them the next time it runs, because the same sequence of IDs will be chosen by the RNG.

**Comment by CendioOssman** [comment]
I don't see any issue resulting from that though?

**Comment by dcommander** [comment]
Seriously?  You can't even acknowledge that you need to seed your random number generator?  Obviously the intent is for you to be generating *random* screen IDs.  Your Java viewer does precisely that.  Your C viewer doesn't.  It if did, then your C viewer would work properly with your own server as well as other servers, including mine, that implement ID matching.  I've explained thoroughly above how this issue is creating an incompatibility with servers, including mine, that do screen ID matching.  This may not be causing any obvious problems with the TigerVNC Server yet, but that doesn't mean it won't.  Long and the short of it:  your intent is to generate random screen IDs.  By not doing that, you are re-using the same screen IDs for different screens, which is an abuse of the protocol.  Let's make this simple.  Are you intending to generate random screen IDs?  Yes?  Then seed your frickin' RNG!  It's not rock science.

**Comment by CendioOssman** [comment]
So far I've only seen you describe a bug in your server because of a misunderstanding of the specification. Seeding the RNG would make that bug less likely to trigger, but it wouldn't actually avoid the bug. You might still get an id that your server secretly remembers from a previous client.

(Using random numbers was merely to avoid anyone thinking that you have to use any kind of well ordered numbers. It is not the mechanism that avoids conflicts.)

**Comment by hildred** [comment]
Not seeding the rng can also cause crypto  failures.

**Comment by dcommander** [comment]
I acknowledge that my server needs to forget the screen IDs of disabled RANDR outputs, and that at least works around the immediate problem, but I also strongly suspect that there are scenarios whereby a similar problem could be reproduced with the TigerVNC Server, because it's matching IDs as well, and it also isn't seeding its RNG.  I just think seeding the RNG is the right thing to do in general, but since I've worked around it, it's not my problem anymore.
